scipy/optimize/_zeros_py.py

```diff
@@ -1534,151 +1533,119 @@ def _chandrupatla(f, x0, x1, *, args=(), xrtol=_xtol,
-
-    # Initialization
-    b = x0
-    a = c = x1
-    fb = f_wrapped(b, *args)
-    fa = fc = f_wrapped(a, *args)
-    t = 0.5
-    iterations = 0
+    def check_convergence(x1, f1, x2, f2, flag, eps_r=1e-5, eps_a=1e-10):
+        # See Section 4
+        xmin, fmin = x2.copy(), f2.copy()
+        imin = np.abs(f1) < np.abs(f2)
+        xmin[imin], fmin[imin] = x1[imin], f1[imin]
 
-    a, b, fa, fb = np.broadcast_arrays(a, b, fa, fb)
-    intermediate_shape = a.shape
-    a, b, fa, fb = np.atleast_1d(a, b, fa, fb)
-    shape = a.shape
+        tol = 2 * eps_r * abs(xmin) + 0.5 * eps_a
+        itol = 0.5 * abs(x2 - x1) < tol
+        itol |= np.abs(fmin) < np.finfo(np.float64).smallest_normal
+        flag[itol] = _ECONVERGED
 
-    iterations = np.zeros_like(a)
+        j = (np.sign(f1) == np.sign(f2)) & ~itol
+        xmin[j], fmin[j], itol[j], flag[
+            j] = np.nan, np.nan, True, _ESIGNERR  # sign error
 
-    # flag to check state of convergence
-    flag = np.full(shape, "in_progress")
+        j = np.isfinite(x1) & np.isfinite(x2) & np.isfinite(f1) & np.isfinite(
+            f2)
+        itol[~j], flag[~j] = True, _EVALUEERR  # value error
 
-    funcalls = np.full(np.shape(flag), 0)
-    terminate = False
+        return xmin, fmin, np.all(itol), flag
 
-    while iterations.any() < maxiter:
-        # use t to linearly interpolate between a and b,
-        # and evaluate this function as our newest estimate xt
-        xt = a + t * (b - a)
-        ft = f_wrapped(xt, *args)
-        funcalls += 1
+    def _wrapper(f):
+        def wrapped(x):
+            wrapped.fevals += 1
+            return f(x, *args)
 
-        # update our history of the last few points so that
-        # - a is the newest estimate (we're going to update it from xt)
-        # - c and b get the preceding two estimates
-        # - a and b maintain opposite signs for f(a) and f(b)
-        samesign = np.sign(ft) == np.sign(fa)
-        c = np.choose(samesign, [b, a])
-        b = np.choose(samesign, [a, b])
-        fc = np.choose(samesign, [fb, fa])
-        fb = np.choose(samesign, [fa, fb])
-        a = xt
-        fa = ft
-
-        # set r so that f(r) is the minimum magnitude of f(a) and f(b)
-        fa_is_smaller = np.abs(fa) < np.abs(fb)
-        r = np.choose(fa_is_smaller, [b, a])
-        fm = np.choose(fa_is_smaller, [fb, fa])
-
-        tol = 2 * xrtol * np.abs(r) + xatol
-        tlim = tol / np.abs(b - a)
-
-        terminate = termination_function()
-
-        # check convergence here, update flag if so
-        if np.all(terminate):
-            flag.fill(_status_message['success'])
-            break
+        wrapped.fevals = 0
+        return wrapped
+
+    # Input validation/standardization
+    func = _wrapper(func)
+
+    x1, x2 = np.broadcast_arrays(x1, x2)
+    xt = np.result_type(x1.dtype, x2.dtype, np.float16)  # at least float16
+    f1, f2 = func(x1), func(x2)
+    ft = np.result_type(f1.dtype, f2.dtype, np.float16)
```

Comment:
Default tolerance should depend on dtype, perhaps. Tolerance can't be achieved with `float16`.

Reply:
Need to rethink the broadcasting and data type enforcing part, too.